### ISSUE TITLE
Missing BrokerAppEntitlementPolicyRule

### ISSUE THREAD

**Issue by Spontrel** [ryancbutler/XDReplicate#28]
## Expected Behavior
  
Import corrupts target Delivery Group
  
Not sure how (at this time) the script is malfunctioning.  It looks correct, but it still corrupts 

## Current Behavior

Process...do a full export (site Primary).  I have 8 delivery groups.  Prior to the import, I delete all applications and all application groups (site Secondary).  All Machine Catalogs and Delivery Groups still exist prior to the import.  After import is complete, no apps launch from secondary site, but there is no distinguishable difference in the GUI.  I opened a case with Citrix and was told the DG is corrupt.  They instructed me to create a new one, then add all the apps to it, and then the apps launch.

This did not occur with the previous non-module version of the script.  I thought it was an odd occurrence the first time this happened, but just did a new export import to my secondary and third sites, and both were corrupted.

## Possible Solution
  Is it possible you are importing GUIDs or UIDS that are not properly cross-referenced?

Testing method:  (the same app is in all 3 sites)  Turn maintenance mode on the DGs for the 2 of 3 sites and test them one by one.  Only the primary site works until the DG is recreated and the apps are re-associated.

## Steps to Reproduce (for bugs)

1.
2.
3.
4.

## Your Environment

* Module Version used:
* Citrix XenDesktop Version:

**Comment by ryancbutler** [comment]
This was an issue awhile back that should be fixed but I don't have anything fixing or detecting broken DGs.

Compare the results of this across your sites.
`Get-BrokerAppEntitlementPolicyRule`

If they are different you should be able to delete the DG and re-import to correct.  Chances are the broken DG is missing this.

**Comment by Spontrel** [comment]
I see that our some of our AppEntitlementPolicyRules have Names that do not match DG Names in the GUI.  I have corrected the mismatches.  Also, one of my sites has no AppEntitlementPolicyRule at all even after running the import.

Are you saying to delete the DG before importing and let the script rebuild it?  If so, I would manually need to add the Machine Catalogs to each DG after it was done.

**Comment by ryancbutler** [comment]
The way it works is the appentitlementpolicyrule is only created when it creates a brand new DG.  It doesn't check or edit existing at this time.  Two options to resolve.

1. Delete the DG and recreate with the import.  You will have to re-associate the machines from a machine catalog.
2. Manually create the appentitlement with `New-BrokerAppEntitlementPolicyRule -Name DGNAME -DesktopGroupUid DGUID -IncludedUserFilterEnabled $false`  You will be able to get the DG UID with `Get-BrokerDesktopGroup`.  Name isn't necessarily as important as the UID to link to the DG.

**Comment by Spontrel** [comment]
Ryan, 

Thanks for all the detail!!!

I wrote a short script to build and/or correct the appentitlementpolicyrule. I'll post it Monday as I'm away from the office today. I noticed the policy rule retains old DG names data even after the DG gets a new name and some new DGs don't all have matching policies based on how they were created. This fits directly with your explanation. This is great news. 

It would be a good addition module item to run after the import completes.  This should solve it. 

Thanks again 

Scott

**Comment by Spontrel** [comment]
asnp citrix*

write-host -foregroundcolor Magenta "Collecting Delivery Group data"
$AllDGs = Get-BrokerDesktopGroup

write-host -foregroundcolor Magenta "Collecting Brokery Policy data"
$allBPRs = Get-BrokerAppEntitlementPolicyRule -ErrorAction 0

#check for existance, and create if missing
write-host -foregroundcolor Magenta "Checking for missing Brokery Policies"
foreach ($DG in $AllDGs){
    $dg_uid = $dg.Uid
    $Dgname = $dg.name

    $BPR = $AllBPRs | Where-Object {$_.DesktopGroupUID -eq $dg_uid}

    if ($BPR -eq $null) {write-host -ForegroundColor Yellow "No Policy for DG: $dgname, creating it now"   
        new-BrokerAppEntitlementPolicyRule -name $dgname -DesktopGroupUid $dg_uid -IncludedUserFilterEnabled $false}
    else {write-host -ForegroundColor cyan "Policy found for DG: $dgname"}
}


#check for proper setup

write-host -foregroundcolor Magenta "Re-checking Brokery Policy data"
$allBPRs = Get-BrokerAppEntitlementPolicyRule 

$BPR = $AllBPRs | Where-Object {$_.DesktopGroupUID -eq $dg_uid} #collect all data again

write-host -foregroundcolor Magenta "Checking Broker Policies and correcting if needed"
foreach ($DG in $AllDGs){

    $dg_uid = $dg.Uid
    $Dgname = $dg.name


    $BPR = $AllBPRs | Where-Object {$_.DesktopGroupUID -eq $dg_uid}

    $BPRname = $BPR.name

    if ($BPRname -eq $dgname) {write-host -ForegroundColor Green "$DGname setup correctly"}
    else {write-host -ForegroundColor Yellow "Fixing $dgname"
        $BPR | rename-BrokerAppEntitlementPolicyRule -newname $dg.name}
}

write-host -ForegroundColor Magenta "Delivery Group:"
$AllDGs | Select-Object name , UID | ft
write-host -ForegroundColor Magenta "Entitlement Policy:"
Get-BrokerAppEntitlementPolicyRule | Select-Object name, desktopGroupUID | ft

**Comment by ryancbutler** [comment]
Want to take a stab at adding this as a function via pull request?

**Comment by Spontrel** [comment]
ug...that was a pain to get working last time.  I'll take a stab at it, but it will probably take me a while to decipher you new module design to integrate it.  I will try!

**Comment by Spontrel** [comment]
Upon further investigation, not only does renaming a Delivery Group mess up the  BrokerAppEntitlementPolicyRule, it also breaks two BrokerAccessPolicyRules.  I opened a case with Citrix this morning.  I believe I have identified a bug in 7.15 around this Delivery Group rename thing.  This is the second bug this week.  I also have another case with Citrix escalation on a different bug I found which they have confirmed exists.
